IDAInterface

luca-heltai opened the pull request:
#4704

luca-heltai commented:
@jdannberg here it is. :)

jdannberg commented:
Great, thanks!

davydden commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,314 @@
+   * scalar \f$\alpha\f$ changes whenever the step size or method order
+   * changes.
+   */
+  template<typename VEC=Vector<double> >
```
Elsewhere we call template VectorType

davydden commented on include/deal.II/sundials/copy.h:
```diff
@@ -0,0 +1,71 @@
+#ifdef DEAL_II_WITH_MPI
+
+#ifdef DEAL_II_WITH_TRILINOS
+  void copy(TrilinosWrappers::MPI::Vector &dst, const N_Vector &src);
```
This and below are not documented

davydden commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,314 @@
+     */
+    IDAInterface(const std::string name="",
+                 const MPI_Comm mpi_comm = MPI_COMM_WORLD);
+    /** House cleaning. */
```
Space before and indentation the same way as constructor

davydden commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,314 @@
+    /** House cleaning. */
+    ~IDAInterface();
+
+    /** Declare parameters for this class to function properly. */
```
I don't understand what u mean by "to function properly"? Can u please reprhrase?

davydden replied:
looking at the test, i would list here explicitly all the parameters which will be declared. Otherwise users are forced to dig into the source code to see what is actually happening inside to be able to use this function.
I would take the output of `ParameterHandler::print_parameters()`.

davydden commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,314 @@
+                        const unsigned int step_number)> output_step;
+
+    /**
+     * Evaluate wether the solver should be restarted (for example because the
```
whether

davydden commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,314 @@
+     */
+    void set_functions_to_trigger_an_assert();
+
+    /** Initial time for the DAE.*/
```
Could u please indent it the way we do elsewhere (i.e. The function above)? Same below

davydden commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,314 @@
+    N_Vector diff_id;
+
+#ifdef DEAL_II_WITH_MPI
+    MPI_Comm communicator;
```
Documentation

davydden commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,314 @@
+    /** Output stream */
+    ConditionalOStream pcout;
+
+    unsigned int system_size;
```
Docu

davydden commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,314 @@
+
+    unsigned int system_size;
+
+    unsigned int local_system_size;
```
Docu

bangerth commented:
@jdannberg -- want to practice reviewing patches with this one? ;-)

davydden commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,589 @@
+  {
+    prm.add_parameter("Initial step size",initial_step_size);
+
+    prm.add_parameter("Min step size", min_step_size);
```
since you say `Maximum` elsewhere, please use `Minimum` here.

davydden commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,589 @@
+
+    prm.add_parameter("Final time", final_time);
+
+    prm.add_parameter("Time units between each output", output_period);
```
not being a native speaker, i perceive `units` as units of time (second, hours, etc), thus the description is confusing  IMO. Maybe change to `Time interval between each output`?

davydden commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,589 @@
+                      "in the local error test.");
+
+    prm.add_parameter("Initial condition type", ic_type,
+                      "This is one of the following thress options for the "
```
thress

davydden commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,589 @@
+                      Patterns::Selection("none|use_y_diff|use_y_dot"));
+
+    prm.add_parameter("Initial condition type after restart", reset_type,
+                      "This is one of the following thress options for the "
```
thress

davydden commented:
@luca-heltai please also add a quick test for the new interface, see  https://github.com/dealii/dealii/issues/3230

bangerth commented:
@luca-heltai ?

luca-heltai commented:
I'm still off the grid with only my phone (and my family :-)). As soon as I'm back to Trieste I'll fix all comments, and add some more stuff.

bangerth commented:
OK, no rush. Enjoy the time!

luca-heltai commented:
I think I have addressed all of @davydden requests. If this is looking good, I'll add to sundials directory also the interfaces to the general nonlinear solver (`KINSOL`), general ODE solver (`CVODE`), and general implicit explicit time stepper (`ARKODE`) with respective tests.

bangerth commented on include/deal.II/sundials/copy.h:
```diff
@@ -0,0 +1,74 @@
+//-----------------------------------------------------------
+//
+//    Copyright (C) 2015 by the deal.II authors
```
You may want to adjust the year.

Do we want to stick with the directory name `sundials/`? Or should we go with something more generic, say `odesolvers/`?

luca-heltai replied:
sundials offers also nonlinear solvers and dae solvers (IDA is a dae solver, for example, not an ode solver...). I'd stick to sundials.

davydden replied:
i agree with @luca-heltai . Eventually we may have some more interfaces to other libs which contain multiple types of solvers so i would group headers based on the library name.

bangerth commented on include/deal.II/sundials/copy.h:
```diff
@@ -0,0 +1,74 @@
+#ifdef DEAL_II_WITH_MPI
+
+#ifdef DEAL_II_WITH_TRILINOS
+    void copy(TrilinosWrappers::MPI::Vector &dst, const N_Vector &src);
```
You choose the `vmult` convention of saying `copy(dst,src)` (which mirrors `dst = src`). But that's not what `std::copy` does, for example, nor what most of the rest of the deal.II functions do. Thoughts?

luca-heltai replied:
I kept this in an internal namespace, because ideally we should get rid of all these in favour of `ReadWriteVector` and `VectorSpaceVector` objects. All copy functions should disappear.

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+   *   \end{cases}
+   * \f]
+   *
+   * where \f$y,\dot y\f$ are vectors in \f$\R^n\f$, \f$t\f$ is often the time (but can
```
You don't need to use `\f` (in fact, it is wrong) since we automatically expand `$` to `\f$`.

luca-heltai replied:
I copyed from deal2lkit... there we don't have this auto-magic expansion... :) I'll fix this.

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+   * also be a parametric quantity), and
+   * \f$F:\R\times\R^n\times\R^n\rightarrow\R^n\f$. Such problem is solved
+   * using Newton iteration augmented with a line search global
+   * strategy. The integration method used in ida is the variable-order,
```
ida -> IDA

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+   * the BDF of order \f$q\f$ given by the multistep formula
+   *
+   * \f[
+   *   \sum\limits_{i=0}^q \alpha_{n,i}\,y_{n-i}=h_n\,\dot y_n\, ,
```
`\limit`? (or maybe I just don't know the command?)

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+   * \f[
+   *   G(y_n)\equiv F\left(t_n,y_n,\dfrac{1}{h_n}\sum\limits_{i=0}^q \alpha_{n,i}\,y_{n-i}\right)=0\, .
+   *   \label{eq:nonlinear}
+   * \end{equation}
```
-> `\f]`

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+   *   \label{eq:jacobian}
+   * \f]
+   *
+   * and \f$\alpha = \alpha_{n,0}/h_n\f$. It is worthing metioning that the
```
-> worth

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+     * Those options are set as default value in the ParameterHandler object
+     * `prm`.
+     */
+    virtual void add_parameters(ParameterHandler &prm);
```
maybe `set_parameters()`? You're basically just *overwriting* existing ones, not adding them.

luca-heltai replied:
Actually I'm calling `prm.add_parameter` for each parameter, which first declare the parameter, and then sets it. I  don't know if `add_parameter` was a good choice in the first place... but here is just using the same name (with a plural).

bangerth replied:
Oh, I misread (or didn't read at all) the documentation. I thought you were setting the parameters of the current object. But you were really setting the parameters of the current class *in the parameter handler object*.

That's the equivalent of the `declare_parameters()` function we describe in the documentation of `ParameterHandler`.

luca-heltai replied:
Not exactly... the `declare_parameters` does not attach to each parameter an action, while this one does.

bangerth replied:
Yes, you're right. It's not equivalent. But maybe reference that function in the documentation and say that it combines the action of `declare_parameters()` and `read_parameters()`.

luca-heltai replied:
In other words, this function also **sets** the parameters whenever `prm` updates its entries.

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+                   VectorType &y,
+                   VectorType &yp,
+                   double h,
+                   bool first_step);
```
can you document what `v` and `yp` are set to (I suspect they are outputs?) and make `h` and `first_step` `const`.

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+     * Store solutions to file.
+     */
+    std::function<void (const double t,
+                        const VectorType &sol,
```
I think that's the wrong documentation -- a user can attach any kind of function here. The function is probably called every time step, right? If so, we use a `boost::signals2::signal` for this in other places.

luca-heltai replied:
Not at every time step, but at fixed times. I'll document this better.

bangerth replied:
The point about the signal is that, just like for linear solvers, one may want to attach one or more listeners to this signal. That's possible with a signal, but not with a `std::function`.

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+   * @author Luca Heltai, Alberto Sartori, 2017.
+   */
+  template<typename VectorType=Vector<double> >
+  class IDAInterface
```
Would just calling this class `IDA` be better?

Also, I think having an example in the documentation would be really useful. Can you cull something from one of the tests you already have?

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+     * called when the simulation start and when the user return true to a call
+     * to solver_should_restart()
+     */
+    void reset_dae(const double t,
```
maybe just call it `reset`?

bangerth commented on include/deal.II/sundials/ida_interface.h:
```diff
@@ -0,0 +1,476 @@
+     * Return a vector whose component are 1 if the corresponding
+     * dof is differential, 0 if algebraic.
+     */
+    std::function<VectorType&()> differential_components;
```
If you only need zeros or ones, why not let the function return a `std::vector<bool>`? The fact that *internally* IDE wants a different format shouldn't dictate using a better interface as we can always do the conversion internally.

luca-heltai replied:
Maybe even better: we could return an IndexSet.

bangerth replied:
Yes, definitely better!

bangerth commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,602 @@
+
+      std::shared_ptr<VectorType> src_yy = solver.create_new_vector();
+      std::shared_ptr<VectorType> src_yp = solver.create_new_vector();
+      std::shared_ptr<VectorType> residual = solver.create_new_vector();
```
I think you want `std::unique_ptr` here.

bangerth commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,602 @@
+      IDAInterface<VectorType> &solver = *static_cast<IDAInterface<VectorType> *>(IDA_mem->ida_user_data);
+
+      std::shared_ptr<VectorType> src_yy = solver.create_new_vector();
+      std::shared_ptr<VectorType> src_yp = solver.create_new_vector();
```
same here and probably below

bangerth commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,602 @@
+  template <typename VectorType>
+  void IDAInterface<VectorType>::add_parameters(ParameterHandler &prm)
+  {
+    prm.add_parameter("Initial step size",initial_step_size);
```
:-)

bangerth commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,602 @@
+    use_local_tolerances(use_local_tolerances),
+    ida_mem(nullptr),
+    communicator(Utilities::MPI::duplicate_communicator(mpi_comm)),
+    pcout(std::cout, Utilities::MPI::this_mpi_process(mpi_comm)==0)
```
Could we do this differently, for example via a signal?

bangerth commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,602 @@
+        next_time += output_period;
+        if (verbose)
+          {
+            pcout << " "//"\r"
```
remove the commented out part

bangerth commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,602 @@
+                  << std::setw(5) << next_time
+                  << std::endl;
+          }
+        status = IDASolve(ida_mem, next_time, &t, yy, yp, IDA_NORMAL);
```
can you add a check of the return value here?

bangerth commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,602 @@
+        status = IDASolve(ida_mem, next_time, &t, yy, yp, IDA_NORMAL);
+
+        status = IDAGetLastStep(ida_mem, &h);
+        AssertThrow(status == 0, ExcMessage("Error in IDA Solver"));
```
I think it would be nice to output the error code. I think there is already something like `AssertMPI` or `Assert(PETSc)` that you could steal for here.

bangerth commented on source/sundials/ida_interface.cc:
```diff
@@ -0,0 +1,602 @@
+            // double frac = 0;
+            int k = 0;
+            IDAGetLastOrder(ida_mem, &k);
+            // frac = std::pow((double)k,2.);
```
remove the commented out part

luca-heltai commented:
I think this is ready for review. As soon as I have some more time, I'll add more of the functionality provided in IDA to this class (with additional tests).

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+{
+
+  /**
+   * Interface to SUNDIALS IDA (Implicit Differential-Algebraic) solver.
```
maybe introduce abbreviation here:`Implicit Differential-Algebraic (IDA)` and then you can use it below?

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+        const bool &use_local_tolerances = false);
+
+    /**
+     * House cleaning.
```
for consistency with other classes, i would write `Destructor`.

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+    virtual void add_parameters(ParameterHandler &prm);
+
+    /**
+     * Evolve. This function returns the final number of steps.
```
maybe `Evolve -> Integrate differential-dlgebraic equations`?

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+     * called when the simulation start and when the user returns true to a
+     * call to solver_should_restart().
+     *
+     * By default solver_should_restart returns false. If the user needs to
```
solver_should_restart()

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+    /**
+     * Minimum step size.
+     */
+    double min_step_size;
```
for consistency with the two above `minimum_step_size`?

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+    /**
+     * Absolute error tolerance for adaptive time stepping.
+     */
+    double abs_tol;
```
`absolute_tolerance`

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+    /**
+     * Relative error tolerance for adaptive time stepping.
+     */
+    double rel_tol;
```
`relative_tolerance`

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+    /**
+     * Maximum order of BDF.
+     */
+    unsigned int max_order;
```
`maximum_order`

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+        const unsigned int &max_order = 5,
+        const double &output_period = .1,
+        const bool &ignore_algebraic_terms_for_errors = true,
+        const std::string &ic_type = "none",
```
i would this and `reset_type`to enumerators and only internally switch to `std::string` to provide the correct input to Sundials. Same for class private variables.

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+     * Output stream. If run in parallel, only processor zero will
+     * output verbose information about time steps.
+     */
+    ConditionalOStream pcout;
```
@dealii/dealii if I understand correctly we are moving away form writing anything from dealii's classes and instead provide functions for optional output

davydden commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+     * Those options are set as default value in the ParameterHandler object
+     * `prm`.
+     */
+    virtual void add_parameters(ParameterHandler &prm);
```
an alternative is to move those into an auxiliary class `AdditionalParameters` which can declare and read things from ParameterHandler. Then you can easily extend the number of parameters later without breaking constructor interface. 

What I find more important is that you can then make all private members `const`.

bangerth commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+   *
+   * That is $F(y', y, t) = y' + A y = 0 $
+   *
+   * where
```
drop the empty line

bangerth commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+   * \end{matrix}
+   * \f]
+   *
+   * and $y(0)=(0, k)$, $y'(0) = (k, 0)$
```
same here.

period at the end of the sentence

bangerth commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+   * and $y(0)=(0, k)$, $y'(0) = (k, 0)$
+   *
+   * The exact solution is $y_0(t) = \sin(k t)$, $y_1(t) = y_0'(t) = k \cos(k t)$,
+   * $y_1'(t) = -k^2 \sin(k t)$
```
period. same in the following line

bangerth commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+    std::function<int(const double t,
+                      const VectorType &y,
+                      const VectorType &y_dot,
+                      const double alpha)> setup_jacobian;
```
since the function only takes const arguments, how does it actually setup a jacobian? Can you say when this function is called, and where the jacobian it sets up is stored?

luca-heltai replied:
it is the user's responsability to decide where to store the Jacobian. This class is agnostic on what the user wants to do, and does not enforce it. I'll document this a little better. The user should make sure that they have a way of solving for `J = alpha dF/dy_dot + dF/dy` after a call to this function is made. If they want to store a matrix, or setup a matrix free computation, it is their responsability, and this class should not care...

luca-heltai replied:
This is now documented.

bangerth commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+     * - <0: Unrecoverable error the computation will be aborted and an assertion
+     *       will be thrown.
+     */
+    std::function<int(const VectorType &rhs, VectorType &dst)> solve_jacobian_system;
```
can you say what the function arguments mean, and that the function is supposed to compute `dst = J^{-1} rhs`?

luca-heltai replied:
done. I don't know why it does not show up here.

bangerth commented on include/deal.II/sundials/ida.h:
```diff
@@ -0,0 +1,653 @@
+    std::function<void (const double t,
+                        const VectorType &sol,
+                        const VectorType &sol_dot,
+                        const unsigned int step_number)> output_step;
```
I think you wanted to update the description (and maybe name) to state when the function will be called and what it is supposed to do.

luca-heltai replied:
done.

luca-heltai commented:
I've rebased and squashed to a smaller number of commits. Not as good as @drwells git-history (from his commits it looks like he always get things right the first time! :) ...) but this should be small enough. 

I think we are now ready.

bangerth commented:
OK to merge once the tester is happy.

/run-tests

bangerth commented:
In reference to #5027.
